Plugin for cloud auto-join

kwohlfahrt opened the pull request:
This PR adds a new type of plugin, for cluster auto-join. Currently, only environments supported by `go-discover` can be used for auto-join, and `go-discover` is considered feature-complete by upstream, and won't support any new environments. Initially, the only plugin implemented is one that wraps the existing `go-discover` functionality.

Also, many cloud provider SDKs are linked into OpenBao just to support Raft auto-join.

As discussed in the issue, this is implemented with a plugin instead of an external command for two reasons - better typing in the interfaces, and to make it easier to ship built-in supported methods.

This looks a bit different to the normal plugin patterns (which use `PluginCatalog`), because IIUC we don't have an unsealed core to use when we are still in the auto-join phase. I'm not sure if I'm using the internal utils idiomatically here, so would appreciate review comments on this specifically.

# TODO

- [x] Tests
- [x] Figure out the linter complaint
- [x] Start plugins once, instead of once per auto-join stanza (?)
- [x] Convert old auto-join into use of discover plugin, and emit deprecation warning (?)

Resolves #1293

kwohlfahrt committed:
Add changelog

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt commented:
@cipherboy since you were discussing the original issue with me, could you take a look at the PR in its current state? It's not ready for a detailed review, but I have a few questions/TODOs in the original PR message that I'd like to get your opinion on before continuing.

kwohlfahrt committed:
Tests for join functionality

Includes config parsing and auto-join with plugin test.

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt committed:
Allow running plugins without a wrapper defined

This is making me think perhaps the pluginutil is not correct for the
join plugin, but tests pass with this.

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt committed:
Initialize join plugins once only

We may want to add a Setup() method as well, to allow for expensive
one-off initialization and configuring of each plugin. But for now, this
can also be done via env & args. This is probably more relevant if we
want to multiplex plugins.

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt committed:
Replace legacy discover calls with plugin

Also make it an error to replace built-in plugins. This way we can
safely rely on the builting discover plugin existing.

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

satoqz reviewed (commented):
I've been getting deeper into the plugin system by working on https://github.com/openbao/openbao/pull/2459 where I'm also adding a new plugin type that will not able to use the plugin catalog (Which I think is fine).

Long story short, that's enabled me to finally give this a somewhat educated look; I think the approach taken here is reasonable :)

One change that I'd request is to reuse the plugin configuration infrastructure that is now available via https://openbao.org/docs/rfcs/config-plugins, so we can manage join plugins using the same stanzas & OCI capabilities that other plugin types now have.

I need to have another look at the Raft changes but wanted to leave a comment nevertheless.

kwohlfahrt committed:
Split plugin catalog and KMS wrapping

The local plugin catalog will also be used for auto-join, which also
needs a plugin catalog that works when sealed.

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt committed:
Move join plugins to shared catalog

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

satoqz reviewed (commented):
Here's another pass @kwohlfahrt; nothing major, just small things to clean up.

kwohlfahrt committed:
Apply suggestions from code review

Co-authored-by: Jonas Köhnen <mail@satoqz.net>
Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt committed:
Mirror api/plugin_types.go

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt committed:
Fix plugin cookie name

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt committed:
Make catalog logger public

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt commented on plugins/join/static/static-join-plugin/main.go:
```diff
@@ -0,0 +1,20 @@
+)
+
+func main() {
+	logger := hclog.Default()
```
> Thoughts on propagating the logger to the plugin server like I did here?

Sounds good, the more consistent the better IMO.

kwohlfahrt replied:
Wait, what is the intended way to pass the logger through to the actual plugin code? In the linked section above, the logger is just passed to `plugin.ServeConfig`, how does it get to the actual plugin code?

satoqz replied:
My bad, didn't set this up for KMS plugins yet because there was no API to pass loggers to `Wrapper` anyway.

Backend plugins are the better reference, the logger is propagated like so:

https://github.com/openbao/openbao/blob/ef9fb04a52d744e53853ec4232e882294225ac01/sdk/plugin/serve.go#L33-L68

And later passed to the plugin's factory here:

https://github.com/openbao/openbao/blob/ef9fb04a52d744e53853ec4232e882294225ac01/sdk/plugin/grpc_backend_server.go#L78-L117

...so coming back to join plugins, I suppose we'd just pass it to the one-off factory right away.

Might be smart to pass a

```go
type JoinConfig struct {
  Logger log.Logger
}
```

to the factory for now to ensure future space for anything we may want later on.

kwohlfahrt replied:
OK, I think I've followed the pattern from the backend plugin. For the AWS backend at least, the ability to pass in a logger isn't actually used though:

https://github.com/openbao/openbao-plugins/blob/c88c2889e3e70b9c609887c115421dd7744cd743/secrets/aws/cmd/main.go#L23-L28

satoqz replied:
@kwohlfahrt hmm I don't think it matters much with the backend plugins because there's almost nothing inside of `ServeMultiplex` can actually fail. I would assume though that for the logging call referenced above to forward to the client correctly, it'd need to be a `JSONFormat=true` logger, not a standard logger.

This does make me think that it's probably better to call the join plugin factory not inside `Serve()`, but inside `GRPCServer()`. `Serve()` can then have no return value at all, and any error produced by the factory will be forwarded after already establishing a gRPC connection, with the correct logging settings by go-plugin.

Edit: Logs are forwarded over stdout, but it probably makes sense to actually handshake/connect before we even call the factory and see if that fails.

satoqz replied:
@kwohlfahrt here's a quick patch based on that idea:

```diff
diff --git a/helper/joinplugin/join.go b/helper/joinplugin/join.go
index 6e5074b81f..d92dc685f4 100644
--- a/helper/joinplugin/join.go
+++ b/helper/joinplugin/join.go
@@ -44,7 +44,9 @@ func (c *Catalog) NewJoin(name string) (joinplugin.Join, bool, error) {
 	}
 	if !ok {
 		if factory, ok := builtins[name]; ok {
-			plugin, err := factory(joinplugin.JoinConfig{c.Logger})
+			plugin, err := factory(&joinplugin.JoinConfig{
+				Logger: c.Logger,
+			})
 			return plugin, true, err
 		}
 		return nil, false, fmt.Errorf("unknown join: %s", name)
diff --git a/plugins/join/discover/discover-join-plugin/main.go b/plugins/join/discover/discover-join-plugin/main.go
index 047a44c947..a1cbd9ac7d 100644
--- a/plugins/join/discover/discover-join-plugin/main.go
+++ b/plugins/join/discover/discover-join-plugin/main.go
@@ -1,24 +1,12 @@
 package main
 
 import (
-	"os"
-
-	"github.com/hashicorp/go-hclog"
 	"github.com/openbao/openbao/plugins/join/discover"
 	"github.com/openbao/openbao/sdk/v2/joinplugin"
 )
 
 func main() {
-	logger := hclog.New(&hclog.LoggerOptions{
-		Level:      hclog.Info,
-		JSONFormat: true,
-	})
-
-	if err := joinplugin.Serve(joinplugin.ServeOpts{
+	joinplugin.Serve(joinplugin.ServeOpts{
 		Factory: discover.Factory,
-		Logger:  logger,
-	}); err != nil {
-		logger.Error(err.Error())
-		os.Exit(1)
-	}
+	})
 }
diff --git a/plugins/join/discover/discover.go b/plugins/join/discover/discover.go
index 3fb3bcb23c..44be0650a2 100644
--- a/plugins/join/discover/discover.go
+++ b/plugins/join/discover/discover.go
@@ -12,7 +12,7 @@ import (
 	"github.com/openbao/openbao/sdk/v2/joinplugin"
 )
 
-func Factory(cfg joinplugin.JoinConfig) (joinplugin.Join, error) {
+func Factory(cfg *joinplugin.JoinConfig) (joinplugin.Join, error) {
 	providers := make(map[string]discover.Provider)
 	for k, v := range discover.Providers {
 		providers[k] = v
diff --git a/plugins/join/static/static-join-plugin/main.go b/plugins/join/static/static-join-plugin/main.go
index fe9c84ae2d..a5fac27ab7 100644
--- a/plugins/join/static/static-join-plugin/main.go
+++ b/plugins/join/static/static-join-plugin/main.go
@@ -1,24 +1,12 @@
 package main
 
 import (
-	"os"
-
-	"github.com/hashicorp/go-hclog"
 	"github.com/openbao/openbao/plugins/join/static"
 	"github.com/openbao/openbao/sdk/v2/joinplugin"
 )
 
 func main() {
-	logger := hclog.New(&hclog.LoggerOptions{
-		Level:      hclog.Info,
-		JSONFormat: true,
-	})
-
-	if err := joinplugin.Serve(joinplugin.ServeOpts{
+	joinplugin.Serve(joinplugin.ServeOpts{
 		Factory: static.Factory,
-		Logger:  logger,
-	}); err != nil {
-		logger.Error(err.Error())
-		os.Exit(1)
-	}
+	})
 }
diff --git a/plugins/join/static/static.go b/plugins/join/static/static.go
index d2b47b5ec4..f06a81e7fe 100644
--- a/plugins/join/static/static.go
+++ b/plugins/join/static/static.go
@@ -11,7 +11,7 @@ import (
 	"github.com/openbao/openbao/sdk/v2/joinplugin"
 )
 
-func Factory(cfg joinplugin.JoinConfig) (joinplugin.Join, error) {
+func Factory(cfg *joinplugin.JoinConfig) (joinplugin.Join, error) {
 	return &Static{logger: cfg.Logger}, nil
 }
 
diff --git a/sdk/joinplugin/join.go b/sdk/joinplugin/join.go
index c31e7d7997..8444c9792c 100644
--- a/sdk/joinplugin/join.go
+++ b/sdk/joinplugin/join.go
@@ -23,14 +23,14 @@ type Join interface {
 type joinPlugin struct {
 	plugin.NetRPCUnsupportedPlugin
 
-	impl Join
+	factory func() (Join, error)
 }
 
 type JoinConfig struct {
 	Logger hclog.Logger
 }
 
-type Factory func(cfg JoinConfig) (Join, error)
+type Factory func(*JoinConfig) (Join, error)
 
 type gRPCClient struct {
 	pb.JoinClient
@@ -94,7 +94,11 @@ func (j *joinPlugin) GRPCClient(ctx context.Context, b *plugin.GRPCBroker, c *gr
 }
 
 func (j *joinPlugin) GRPCServer(b *plugin.GRPCBroker, s *grpc.Server) error {
-	pb.RegisterJoinServer(s, &gRPCServer{impl: j.impl})
+	impl, err := j.factory()
+	if err != nil {
+		return err
+	}
+	pb.RegisterJoinServer(s, &gRPCServer{impl: impl})
 	return nil
 }
 
diff --git a/sdk/joinplugin/serve.go b/sdk/joinplugin/serve.go
index 0265c96597..f508587c91 100644
--- a/sdk/joinplugin/serve.go
+++ b/sdk/joinplugin/serve.go
@@ -10,7 +10,7 @@ type ServeOpts struct {
 	Logger  hclog.Logger
 }
 
-func Serve(opts ServeOpts) error {
+func Serve(opts ServeOpts) {
 	logger := opts.Logger
 	if logger == nil {
 		logger = hclog.New(&hclog.LoggerOptions{
@@ -19,19 +19,18 @@ func Serve(opts ServeOpts) error {
 		})
 	}
 
-	impl, err := opts.Factory(JoinConfig{logger})
-	if err != nil {
-		// Don't log returned error, caller will handle it,
-		return err
+	factory := func() (Join, error) {
+		return opts.Factory(&JoinConfig{
+			Logger: logger,
+		})
 	}
 
 	plugin.Serve(&plugin.ServeConfig{
 		HandshakeConfig: HandshakeConfig,
 		VersionedPlugins: map[int]plugin.PluginSet{
-			1: {"join": &joinPlugin{impl: impl}},
+			1: {"join": &joinPlugin{factory: factory}},
 		},
 		GRPCServer: plugin.DefaultGRPCServer,
 		Logger:     logger,
 	})
-	return nil
 }
```

kwohlfahrt committed:
Rename (Join|KMS)Catalog to Catalog

The type is implicit from the package name.

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt commented on helper/joinplugin/join.go:
```diff
@@ -0,0 +1,73 @@
+	"github.com/openbao/openbao/sdk/v2/joinplugin"
+)
+
+type JoinCatalog struct {
```
Yes, I prefer the shorter, `fooplugin.FooCatalog` is redundant.

kwohlfahrt commented on helper/pluginutil/catalog/catalog.go:
```diff
@@ -72,23 +80,35 @@ func NewCatalog(logger hclog.Logger, config *server.Config) (*Catalog, error) {
 
-func (c *Catalog) getClientLocked(name string) (*client, bool, error) {
+func (c *Catalog) CloseAll() {
+}
```
Whoops, that was left over from some early prototyping.

kwohlfahrt commented on helper/pluginutil/catalog/catalog.go:
```diff
@@ -72,23 +80,35 @@ func NewCatalog(logger hclog.Logger, config *server.Config) (*Catalog, error) {
+	if stored, ok := c.clients[client.name]; ok && stored == client {
+		delete(c.clients, client.name)
+	}
+}
```
No, this was left over from the `CloseAll()` prototyping above. Have inlined it.

kwohlfahrt committed:
Remove CloseAll

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt committed:
Add error for runtime plugin catalog

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt committed:
Revert "Allow running plugins without a wrapper defined"

This reverts commit be7116ee04ccafb13bb01ff1bc416e474ff1e283.

This is no longer necessary with the join catalog.

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt committed:
Remove TODO comment

This question was answered in PR review.

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt commented on sdk/joinplugin/join.go:
```diff
@@ -0,0 +1,100 @@
+	Impl Join
+}
+
+// TODO: Should this take a config/context, like logical.Factory?
```
I was thinking about whether we should support plugin multiplexing as well? IIUC currently, if the same plugin type is instantiated twice, only one client process will be launched, so all of these plugin instantiations will share the same state.

satoqz replied:
I'm not sure yet what the value in serving multiple `Join` instances at once would be for join plugins. Based on the current design, a single instance of `Join` should be enough to handle several configurations passed via `Candidates()` while the `Join` itself remains largely stateless, no?

kwohlfahrt replied:
I also don't have an immediate use-case, if it can be added later in a backwards compatible way (like the `Setup` method), let's do it when there is demand.

satoqz commented on sdk/joinplugin/pb/join.proto:
```diff
@@ -0,0 +1,25 @@
+message Empty {}
+
+message CandidateArgs {
+    map<string, string> config = 1;
```
I wonder if we should use https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/struct.proto here, so we can pass the JSON-serializable subset of `map[string]any` and then have plugins parse that with `mapstructure.WeakDecode`. This gets you much nicer array parsing, for example.

Available via:

```proto
import "google/protobuf/struct.proto";

message CandidateArgs {
    google.protobuf.Struct config = 1;
}
```

satoqz replied:
For an almost out-of-context reference, I had similar plans with https://github.com/Ki-Reply-GmbH/go-kms-wrapping/blob/cf7ab09a9c0d714f3442a404ad394f9896fb686e/kms/kms.go#L120-L127 which is part of the proposed future addition for KMS plugins. Existing plugin interfaces (such as `Wrapper`) take `map[string]string` for configuration, but I found that very cumbersome to work with; it probably contributes to 1000+ lines of parsing boilerplate across all of go-kms-wrapping 😄

kwohlfahrt replied:
I like this idea, but if you know how to implement it off-hand, I'd appreciate if you could push/PR an example for the static plugin config (especially how the decoding should work).

Otherwise I'll figure it out, but it might take another couple of days, I only have a few hours in the evening to work on this.

satoqz replied:
This is the gist of how you'd parse it in the plugin code:

https://github.com/Ki-Reply-GmbH/go-kms-wrapping/blob/f7fdf585923421cd0cb324baaa899b561f025610/kms/transit/transit.go#L51-L70

And here's how you'd convert between `map[string]any` and the `google.protobuf.Struct` type in Go:

Client: https://github.com/Ki-Reply-GmbH/go-kms-wrapping/blob/5c010f42dec73eec66d3bedd75eb1fe1b403bb8c/plugin/kms_client.go#L49-L60
Server: https://github.com/Ki-Reply-GmbH/go-kms-wrapping/blob/5c010f42dec73eec66d3bedd75eb1fe1b403bb8c/plugin/kms_server.go#L80

If time runs low, just let me know and I'll try to get this over the finish line (though I think we're very close). I've optimistically added this to the v2.6-beta milestone for now, but there's no hard requirement at all to ship until then.

kwohlfahrt committed:
Remove unnecessarily public fields in joinplugin

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt commented on sdk/joinplugin/join.go:
```diff
@@ -0,0 +1,100 @@
+	plugin.NetRPCUnsupportedPlugin
+
+	Impl Join
+}
```
Looks like no (and also `gRPCServer.Impl`).

kwohlfahrt committed:
Pass through logger

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt committed:
Pass logger into join plugin factory

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

satoqz commented on sdk/joinplugin/serve.go:
```diff
@@ -0,0 +1,37 @@
+	Logger  hclog.Logger
+}
+
+func Serve(opts ServeOpts) error {
```
```suggestion
func Serve(opts *ServeOpts) error {
```

Let's make this take a reference just out of consistency with other plugins.

kwohlfahrt committed:
Delay join plugin factory

This way, we can postpone all errors until we have an actual logger.

Co-authored-by: Jonas Köhnen <mail@satoqz.net>
Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt committed:
Allow structured config in join plugin

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt commented:
@satoqz OK, I think that's everything from the review - could you take another look please?

satoqz commented on plugins/join/static/static.go:
```diff
@@ -0,0 +1,54 @@
+
+	candidates := make([]joinplugin.Addr, 0, len(cfg.Addresses))
+	for _, rawUrl := range cfg.Addresses {
+		url, err := url.Parse(rawUrl)
+		if err != nil {
+			d.logger.Warn("failed to parse URL", rawUrl, err)
+		}
+		host := url.Hostname()
+		portStr := url.Port()
+		if portStr == "" {
+			portStr = "8200"
+		}
+		port, err := strconv.ParseUint(portStr, 10, 16)
+		if err != nil {
+			d.logger.Warn("failed to parse port", portStr)
+		}
```
Should the error branches return the error rather than logging?

kwohlfahrt replied:
Hm, they should at least `continue`. Depends whether we want to consider the remaining addresses if one of them is invalid. Just returning the error is simpler, and I don't think anyone will complain about an invalid config not working.

satoqz replied:
Yeah I'd just return the error on the spot.

satoqz commented on plugins/join/static/static.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+type staticConfig struct {
+	Addresses []string
```
```suggestion
	Addresses []string `mapstructure:"addresses"`
```

I think it's good to be explicit about the expected name (in the discover plugin, too).

satoqz commented on physical/raft/raft_test.go:
```diff
@@ -283,6 +283,45 @@ func TestRaft_ParseNonVoter(t *testing.T) {
+					"config": {
+					  "provider": "aws",
+					  "args": {"foo": "bar"}
+		            }
```
```suggestion
					"config": {
					  "provider": "aws",
					  "args": {"foo": "bar"}
		      }
```

satoqz commented on vault/raft.go:
```diff
@@ -1034,16 +1028,44 @@ func (c *Core) JoinRaftCluster(ctx context.Context, leaderInfos []*raft.LeaderJo
+				for name, plugin := range joinPlugins {
+					err := plugin.Cleanup(ctx)
+					if err != nil {
+						c.logger.Error("error cleaning up join plugin %s: %w", name, err)
```
```suggestion
						c.logger.Error("error cleaning up join plugin", "name", name, "error", err.Error())
```

satoqz commented on vault/raft.go:
```diff
@@ -1054,8 +1076,16 @@ func (c *Core) JoinRaftCluster(ctx context.Context, leaderInfos []*raft.LeaderJo
+			for name, plugin := range joinPlugins {
+				err := plugin.Cleanup(ctx)
+				if err != nil {
+					c.logger.Error("error cleaning up join plugin %s: %w", name, err)
```
```suggestion
					c.logger.Error("error cleaning up join plugin", "name", name, "error", err.Error())
```

satoqz commented on changelog/1556.txt:
````diff
@@ -0,0 +1,3 @@
+```release-note:feature
+storage/raft: auto-join is now a pluggable feature, allowing support for new environments with custom plugins.
````
```suggestion
Add **Auto Join plugins** for Raft to support discovery in arbitrary environments via custom external plugins. 
```

`feature` type release notes take a slightly different format.

satoqz commented on helper/pluginutil/catalog/catalog.go:
```diff
@@ -29,11 +28,14 @@ import (
 // created. Additionally, we get to simplify many things as KMS plugins are
 // declarative only.
```
```suggestion
// Catalog manages long-lived plugin server processes and clients of a
// particular plugin type. It is disjoint from the plugin catalog in core
// and services plugins that must be available before core starts up and are
// configurable via the server configuration only.
```

satoqz commented on helper/pluginutil/catalog/catalog.go:
```diff
@@ -42,7 +44,13 @@ type Catalog struct {
 }
 
 // NewCatalog returns a new KMS plugin catalog.
```
```suggestion
// NewCatalog returns a new Catalog.
```

satoqz commented on physical/raft/raft.go:
```diff
@@ -231,6 +237,12 @@ type LeaderJoinInfo struct {
+	// AutoJoinScript defines an executable and arguments that outputs peer API
+	// addresses (e.g. `https://vault.example.com:8200`) to standard output. If
+	// supplied, Vault will attempt to automatically discover peers in addition
+	// to what can be provided via 'leader_api_addr'.
```
This seems outdated? FWIW we also don't ever refer to OpenBao as Vault.

kwohlfahrt replied:
Yes it is out of date. Sorry about the stray reference, I'll go fix the comment I copied this from too then 😉  https://github.com/openbao/openbao/blob/7d2db749fdc3b4b7df2ef55d385a26a9202c8aab/physical/raft/raft.go#L221-L223

satoqz reviewed (commented):
Final suggestions. @openbao/openbao-org-maintainers do you want to have a look? I think this is in fairly good shape now.

kwohlfahrt committed:
Explicitly label mapstructure fields

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt committed:
Return errors in static plugin

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt committed:
Rephrase join plugin docstring

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt committed:
Apply documentation tweaks from code review

Co-authored-by: Jonas Köhnen <mail@satoqz.net>
Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt committed:
Apply logging suggestions from code review

Co-authored-by: Jonas Köhnen <mail@satoqz.net>
Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt committed:
Whitespace fixes

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

ragnar-howler reviewed (commented):
🤖 **Automated PR Review Complete**

This PR has been scanned for:
- ✅ Common security issues
- ✅ Code style consistency  
- ✅ Test coverage indicators
- ✅ Documentation completeness

**Stats:** +1316 -123 in 26 files

Please address any automated checks above and request human review when ready.

cipherboy commented on helper/kmsplugin/wrapper_test.go:
```diff
@@ -11,11 +11,53 @@ import (
 )
 
+const TestPluginServerEnv = "BAO_TEST_PLUGIN_SERVER"
+
+// TestPluginServer is not an actual test but a hack to reuse the test binary
+// as a plugin binary that is called into by the same test binary from the main
+// test runner process.
```
🤯

cipherboy commented on physical/raft/raft.go:
```diff
@@ -270,6 +281,19 @@ type LeaderJoinInfo struct {
 }
 
+func (info *LeaderJoinInfo) ValidateJoinMethods() error {
+	// TODO: At config loading, convert all join methods (legacy discover and
```
Since we're making a lot of pluginization related changes in v2.7.0, let's add a deprecation warning about `discover` and move it to `openbao-plugins` by then @satoqz @kwohlfahrt.

satoqz replied:
That's what I had in mind, yeah.

satoqz replied:
I've marked this as follow-up work here, together with general documentation: https://github.com/openbao/openbao/issues/2956.

cipherboy reviewed (dismissed):
Code is looking good, though I don't run Raft enough to test live.

Will let @satoqz do the final review, but thank you @kwohlfahrt and appreciate your patience! This is looking rather excellent.

satoqz commented on vault/raft.go:
```diff
@@ -1034,16 +1028,44 @@ func (c *Core) JoinRaftCluster(ctx context.Context, leaderInfos []*raft.LeaderJo
+					if err != nil {
+						c.logger.Error("error cleaning up join plugin", "name", name, "error", err.Error())
+					}
+				}
```
This also needs to run if we fail at `NewJoin` above, and already spawned plugins just in case several are configured. Probably best to pull it out as a `cleanupPlugins` closure and call in all 3 places.

satoqz replied:
Similarly, nothing waits for this Goroutine to eventually complete so we're not sure to actually run the defer.

satoqz replied:
Yeah, this Goroutine always leaks if `join()` never succeeds. `ctx` always being `context.Background()` doesn't help either...

satoqz replied:
This also led me to discover that `/sys/storage/raft/join` allows passing `auto_join` and `retry`, should we support passing `auto_join_plugin` there? See `http/sys_raft.go`.

kwohlfahrt replied:
> This also needs to run if we fail at NewJoin above, and already spawned plugins just in case several are configured. Probably best to pull it out as a cleanupPlugins closure and call in all 3 places.

Done.

> Yeah, this Goroutine always leaks if join() never succeeds. ctx always being context.Background() doesn't help either...

Is this not expected? If the join hasn't succeeded, the aim is to retry forever, no? And as long as we're retrying, we need to keep the plugins alive.

> This also led me to discover that /sys/storage/raft/join allows passing auto_join and retry, should we support passing auto_join_plugin there? See http/sys_raft.go.

Probably? Is that endpoint intended to be the API version of the `retry_join` config? So you `POST` it on a new member instead of specifying `retry_join` in the config file?

There aren't any tests for this endpoint as far as I can see, so I'd have to put those together from scratch.

satoqz replied:
> Is this not expected? If the join hasn't succeeded, the aim is to retry forever, no? And as long as we're retrying, we need to keep the plugins alive.

By leak I meant to say it does not synchronously shut down before `main()` exits, and as a result we won't run `defer`s, which means we leak the plugin process (which go-plugin doesn't attach to the parent fsr). To give an example: https://go.dev/play/p/uEZ6UlynXl0.

> Probably? Is that endpoint intended to be the API version of the retry_join config? So you POST it on a new member instead of specifying retry_join in the config file?

I assume so too (guesswork for me as well). But let's defer this to a follow up since it's a niche detail we can add separately.

kwohlfahrt replied:
Ah I see. Interestingly the playground does run them all about 90% of the time for me, but sometimes it does leak.

Is it a good idea to put a `ctx context.Context` on the core, and then do a `defer ctx.Cancel()` in the main `Run()` method? It should probably be used in more places, lots of `context.Background()` everywhere.

cipherboy replied:
@kwohlfahrt I have a patch locally that makes active context become atomic, but I think that's more narrow in scope than this one. Context on core could be an interesting idea, though I think I'd need to understand how it interacts with the command line and that weird `-exit-on-core-shutdown` flag. Either way, should probably be bound to core lifetime. 

I think that'd be good.

kwohlfahrt replied:
~Actually there is already `core.GetContext()`, which looks like something that can be used here. Let me give that a spin.~

```go
	// This can be used to trigger operations to stop running when Vault is
	// going to be shut down, stepped down, or sealed
	activeContext           context.Context
```

EDIT: Actually nevermind, probably not. We need to join while the core is sealed, so it's not a great fit. I'll just use a new context for the main `Run()` function.

cipherboy replied:
@kwohlfahrt Sadly that's tied to the lifetime of the active/read-enabled standby state, not the entire server process lifetime. It isn't created early enough I don't think either.

satoqz commented:
I wanted to give this a final test run but I'm having trouble defining the config as HCL.

I'm using the following:

```hcl
  retry_join {
    auto_join_plugin = {
      plugin = "static"
      config = {
        addresses = ["http://127.0.0.1:8200"]
      }
    }
  }
```

where the contents of the `retry_join` stanza are parsed in `command/server/config.go` and converted to the following JSON, which is then passed to Raft and deserialized back from JSON:

```json
[
  {
    "auto_join_plugin": [
      {
        "config": [
          {
            "addresses": [
              "http://127.0.0.1:8200"
            ]
          }
        ],
        "plugin": "static"
      }
    ]
  }
]
```

Note how we expect the outer array (this is `[]*LeaderJoinInfo`), but don't expect `auto_join_plugin` to be an array, neither do we expect `config` to be an array.

We can adapt code to expect these, but that's really ugly, especially since this expands to any potentially plugin-defined nested key.

Not immediately sure how to work around this, thoughts?

satoqz commented:
To illustrate, attached is an ugly patch that resolves most of the described HCL problem, but not very nicely. This would result in:
- Allowing several `auto_join_plugin` stanzas per `retry_join` stanza, which seems reasonable.
- Allowing only a single `config` block.
- When join plugins define nested config objects (likely rare, but supported), they must perform the same de-duplication down to a single object if those are the wanted semantics.

<details>

<summary>
Expand patch
</summary>

```diff
diff --git a/physical/raft/raft.go b/physical/raft/raft.go
index 71879061ed..60fcd26f61 100644
--- a/physical/raft/raft.go
+++ b/physical/raft/raft.go
@@ -217,8 +217,9 @@ func (r *RaftBackend) HookInvalidate(hook physical.InvalidateFunc) {
 }
 
 type AutoJoinPlugin struct {
-	Plugin string         `json:"plugin"`
-	Config map[string]any `json:"config"`
+	Plugin    string           `json:"plugin"`
+	ConfigRaw []map[string]any `json:"config"`
+	Config    map[string]any   `json:"-"`
 }
 
 // LeaderJoinInfo contains information required by a node to join itself as a
@@ -237,10 +238,10 @@ type LeaderJoinInfo struct {
 	// auto-join.
 	AutoJoinPort uint `json:"auto_join_port"`
 
-	// AutoJoinPlugin specifies the name and configuration of an existing join
-	// plugin. If supplied, OpenBao will attempt to automatically discover peers
-	// in addition to what can be provided via 'leader_api_addr'.
-	AutoJoinPlugin *AutoJoinPlugin `json:"auto_join_plugin"`
+	// AutoJoinPlugins specify names and configurations of existing join
+	// plugins. If supplied, OpenBao will attempt to automatically discover
+	// peers in addition to what can be provided via 'leader_api_addr'.
+	AutoJoinPlugins []*AutoJoinPlugin `json:"auto_join_plugin"`
 
 	// LeaderAPIAddr is the address of the leader node to connect to
 	LeaderAPIAddr string `json:"leader_api_addr"`
@@ -285,7 +286,7 @@ func (info *LeaderJoinInfo) ValidateJoinMethods() error {
 	// TODO: At config loading, convert all join methods (legacy discover and
 	// static addresses) to plugins. Then, this can be removed.
 	haveJoinMethod := false
-	for _, joinMethodEnabled := range []bool{len(info.AutoJoin) != 0, len(info.LeaderAPIAddr) != 0, info.AutoJoinPlugin != nil} {
+	for _, joinMethodEnabled := range []bool{len(info.AutoJoin) != 0, len(info.LeaderAPIAddr) != 0, len(info.AutoJoinPlugins) != 0} {
 		if haveJoinMethod && joinMethodEnabled {
 			return errors.New("only one of leader_api_addr, auto_join, and auto_join_script may be enabled")
 		}
@@ -325,10 +326,10 @@ func (b *RaftBackend) JoinConfig() ([]*LeaderJoinInfo, error) {
 			}
 			provider := config["provider"]
 			delete(config, "provider")
-			info.AutoJoinPlugin = &AutoJoinPlugin{
+			info.AutoJoinPlugins = []*AutoJoinPlugin{{
 				Plugin: "discover",
 				Config: map[string]any{"provider": provider, "args": map[string]string(config)},
-			}
+			}}
 			info.AutoJoin = ""
 		}
 
@@ -336,6 +337,20 @@ func (b *RaftBackend) JoinConfig() ([]*LeaderJoinInfo, error) {
 			return nil, fmt.Errorf("invalid scheme %q; must either be http or https", info.AutoJoinScheme)
 		}
 
+		for _, plugin := range info.AutoJoinPlugins {
+			// Lower the raw config array into a single config. We always
+			// receive an array of configs from the server config because it is
+			// interpreted as an HCL block, even in JSON mode.
+			switch len(plugin.ConfigRaw) {
+			case 0:
+			case 1:
+				plugin.Config = plugin.ConfigRaw[0]
+			default:
+				return nil, fmt.Errorf("invalid auto_join_plugin config: expected a single 'config' block, got %d", len(plugin.ConfigRaw))
+			}
+			plugin.ConfigRaw = nil
+		}
+
 		info.Retry = true
 		info.TLSConfig, err = b.parseTLSInfo(info)
 		if err != nil {
diff --git a/physical/raft/raft_test.go b/physical/raft/raft_test.go
index 2a01c37b57..7a2f3dbd74 100644
--- a/physical/raft/raft_test.go
+++ b/physical/raft/raft_test.go
@@ -289,13 +289,13 @@ func TestRaft_JoinConfig(t *testing.T) {
 		conf: map[string]string{
 			"retry_join": `[
 				{"auto_join": "provider=aws foo=baz"},
-				{"auto_join_plugin": {
+				{"auto_join_plugin": [{
 					"plugin": "discover",
-					"config": {
+					"config": [{
 						"provider": "aws",
 						"args": {"foo": "bar"}
-					}
-				}}
+					}]
+				}]}
 			]`,
 		},
 	}
@@ -305,17 +305,17 @@ func TestRaft_JoinConfig(t *testing.T) {
 	}
 	expected := []*LeaderJoinInfo{
 		{
-			AutoJoinPlugin: &AutoJoinPlugin{
+			AutoJoinPlugins: []*AutoJoinPlugin{{
 				Plugin: "discover",
 				Config: map[string]any{"provider": "aws", "args": map[string]string{"foo": "baz"}},
-			},
+			}},
 			Retry: true,
 		},
 		{
-			AutoJoinPlugin: &AutoJoinPlugin{
+			AutoJoinPlugins: []*AutoJoinPlugin{{
 				Plugin: "discover",
 				Config: map[string]any{"provider": "aws", "args": map[string]any{"foo": "bar"}},
-			},
+			}},
 			Retry: true,
 		},
 	}
diff --git a/vault/external_tests/raft/raft_test.go b/vault/external_tests/raft/raft_test.go
index d46a05e268..d9e26240f2 100644
--- a/vault/external_tests/raft/raft_test.go
+++ b/vault/external_tests/raft/raft_test.go
@@ -220,12 +220,12 @@ func TestRaft_RetryAutoJoin_Plugin(t *testing.T) {
 
 	leaderInfos := []*raft.LeaderJoinInfo{
 		{
-			AutoJoinPlugin: &raft.AutoJoinPlugin{
+			AutoJoinPlugins: []*raft.AutoJoinPlugin{{
 				Plugin: "static",
 				Config: map[string]any{
 					"addresses": []string{leaderCore.Client.Address()},
 				},
-			},
+			}},
 			TLSConfig: leaderCore.TLSConfig(),
 			Retry:     true,
 		},
diff --git a/vault/raft.go b/vault/raft.go
index f69b752e9d..12a03002e4 100644
--- a/vault/raft.go
+++ b/vault/raft.go
@@ -1035,18 +1035,16 @@ func (c *Core) JoinRaftCluster(ctx context.Context, leaderInfos []*raft.LeaderJo
 
 	joinPlugins := make(map[string]joinsdk.Join, 0)
 	for _, info := range leaderInfos {
-		if info.AutoJoinPlugin == nil {
-			continue
-		}
-		name := info.AutoJoinPlugin.Plugin
-		if _, ok := joinPlugins[name]; ok {
-			continue
-		}
-		plugin, _, err := catalog.NewJoin(name)
-		if err != nil {
-			return false, fmt.Errorf("failed to create join plugin: %w", err)
+		for _, plugin := range info.AutoJoinPlugins {
+			if _, ok := joinPlugins[plugin.Plugin]; ok {
+				continue
+			}
+			join, _, err := catalog.NewJoin(plugin.Plugin)
+			if err != nil {
+				return false, fmt.Errorf("failed to create join plugin: %w", err)
+			}
+			joinPlugins[plugin.Plugin] = join
 		}
-		joinPlugins[name] = plugin
 	}
 
 	if retryFailures {
@@ -1104,17 +1102,19 @@ func (c *Core) raftLeaderInfo(ctx context.Context, joinPlugins map[string]joinsd
 	switch {
 	case leaderInfo.LeaderAPIAddr != "":
 		ret = append(ret, leaderInfo)
-	case leaderInfo.AutoJoinPlugin != nil:
-		join := joinPlugins[leaderInfo.AutoJoinPlugin.Plugin]
-		addrs, err := join.Candidates(ctx, leaderInfo.AutoJoinPlugin.Config)
-		if err != nil {
-			return nil, fmt.Errorf("failed to run join plugin: %w", err)
-		}
+	case len(leaderInfo.AutoJoinPlugins) != 0:
+		for _, plugin := range leaderInfo.AutoJoinPlugins {
+			join := joinPlugins[plugin.Plugin]
+			addrs, err := join.Candidates(ctx, plugin.Config)
+			if err != nil {
+				return nil, fmt.Errorf("failed to run join plugin: %w", err)
+			}
 
-		for _, addr := range addrs {
-			info := *leaderInfo
-			info.LeaderAPIAddr = fmt.Sprintf("%s://%s:%d", addr.Scheme, addr.Host, addr.Port)
-			ret = append(ret, &info)
+			for _, addr := range addrs {
+				info := *leaderInfo
+				info.LeaderAPIAddr = fmt.Sprintf("%s://%s:%d", addr.Scheme, addr.Host, addr.Port)
+				ret = append(ret, &info)
+			}
 		}
 	default:
 		return nil, errors.New("must provide leader address or auto-join metadata")
```

</details>

---

@cipherboy notably, we can't apply the profile system hack we talked about so easily since the storage plugin itself may want to choose between both behaviors. For example, Raft expects multiple `retry_join` stanzas, so we can't eliminate `[]map[string]any` _before_ passing the config to the storage backend, neither can we eliminate it _after_ it is passed to the storage backend because it will see `[]any` instead after roundtripping from JSON.

kwohlfahrt committed:
Clean up join plugins when failing to create

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt committed:
Wait for plugins to be cleaned up

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

kwohlfahrt committed:
First implementation of join plugin

I think I'm wrapping my head around how the plugin system is supposed to
work.

Includes a discover plugin that matches the existing auto-join
functionality, and a static plugin for testing.

Signed-off-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>
